lib/block/blockdownload.js

```diff
@@ -0,0 +1,223 @@
+ */
+async function* generateBlockDownloadTransfer(options) {
+    console.log(`Generating block download transfer parts`);
+    const expectedLength = options.downloadFiles && (options.downloadFiles.length || options.downloadFiles.size);
+
+    let assetCounter = 0;
+    for (const downloadFile of options.downloadFiles) {
+
+        const sourceUrl = new URL(downloadFile.fileUrl);
+        const targetPath = fileUrl(downloadFile.filePath);
+
+        const source = new Asset(sourceUrl, options.headers);
+        const target = new Asset(targetPath);
+
+        const transferAsset = new TransferAsset(source, target, {
+            acceptRanges: true,
+            metadata: new AssetMetadata(downloadFile.filePath, undefined, downloadFile.fileSize)
+        });
+
+        assetCounter++;
+        console.log(`Generated download transfer asset ${assetCounter} of ${expectedLength}`);
+
+        yield transferAsset;
+    }
+
+    console.log(`Generated ${assetCounter} download transfer assets (files to download: ${expectedLength})`);
+}
+
+/**
+ * Provide a single file download process which is allowed to run in parallel
+ */
+
+const DEFAULT_MAX_CONCURRENCY = 8;
+// Default part size is 10mb
+const DEFAULT_PART_SIZE = 10 * 1024 * 1024; 
+
+/**
+ * @typedef {Object} DownloadFile
+ * @property {String} fileUrl AEM url of file to download
+ * @property {String} filePath Path on the local disk where to download
+ * @property {Number} fileSize Size of the file being downloaded
+ */
+
+class BlockDownload extends EventEmitter {
+    /**
+     * @typedef {Object} BlockDownloadOptions
+     * @property {Integer} maxConcurrent Maximum number of concurrent HTTP requests that are allowed (deafult is 8)
+     * @property {DownloadFile[]} downloadFiles Files that will be downloaded.
+     * @property {Boolean} retryEnabled If true, retries will be attempted for failed parts
+     * @property {Number} [preferredPartSize] Preferred part size
+     * @property {Number} [retryCount] Retry count (default is 5)
+     */
+
+    /**
+     * Constructs a new block download (event emitter) instance
+     */
+    constructor(){
+        super();
+        this.errorEvents = null;
+    }
+
+    /**
+     * Create a a block download controller, which emits events
+     * on file (download) start, file (download) progress, file (download) end
+     * and file (download) error.
+     * @returns {TransferController} Transfer controller
+     */
+    createBlockDownloadController() {
+        const controller = new TransferController();
+        this.errorEvents = [];
+
+        controller.on(TransferEvents.CREATE_TRANSFER_PARTS, transferEvent => {
+            console.log("Block download: block download controller starting file download");
+            this.emit("filestart", transferEvent.transferAsset.eventData);
+        });
+
+        controller.on(TransferEvents.JOIN_TRANSFER_PARTS, transferEvent => {
+            this.emit("fileprogress", {
```

Comment:
lets also log `file progress` for observability

Reply:
No, it is too noisy (already in tests).